src/Conversion/ONNXToKrnl/Tensor/LayoutTransform.cpp

```diff
@@ -0,0 +1,84 @@
+
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//===---------- LayoutTransform.cpp - Lowering Layout Transform Op --------===//
+//
+// Copyright 2022 The IBM Research Authors.
+//
+// =============================================================================
+//
+// This file lowers the ONNX Layout Transform Operator to Krnl dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "src/Conversion/ONNXToKrnl/ONNXToKrnlCommon.hpp"
+#include "src/Dialect/ONNX/ShapeInference/ONNXShapeHelper.hpp"
+
+using namespace mlir;
+
+namespace onnx_mlir {
+
+struct ONNXLayoutTransformOpLowering : public ConversionPattern {
+  ONNXLayoutTransformOpLowering(TypeConverter &typeConverter, MLIRContext *ctx)
+      : ConversionPattern(typeConverter,
+            mlir::ONNXLayoutTransformOp::getOperationName(), 1, ctx) {}
+
+  LogicalResult matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const final {
+    ONNXLayoutTransformOpAdaptor operandAdaptor(operands);
+    ONNXLayoutTransformOp LayoutTransformOp =
+        llvm::cast<ONNXLayoutTransformOp>(op);
+    auto loc = op->getLoc();
+
+    // Operands and attributes.
+    Value data = operandAdaptor.data();
+
+    // Convert the input type to MemRefType.
+    Type inConvertedType = typeConverter->convertType(data.getType());
+    assert(inConvertedType && inConvertedType.isa<MemRefType>() &&
+           "Failed to convert type to MemRefType");
+    MemRefType inMemRefType = inConvertedType.cast<MemRefType>();
+    uint64_t rank = inMemRefType.getShape().size();
```

Comment:
Why don't we get rank from outMemRefType? Looking at the loop below, I guess the input and output must have the same rank.

Reply:
Yes, the LayoutTransform operation by definition have inputs and outputs of the same logical rank. The layout, when not the default one, introduce a map which will increase the dimensionality by 1 or 2 (depending on the current layouts) of the MAPPED data. Let me add a comment in the code to stress that point.

Reply:
But it looks like we don't have the mapped data at this stage unless we normalize memrefs before this lowering.
Anyway, we can get rank from dataBounds also.

Reply:
done